ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/src/dto/AssignmentRequestDto.ts

```diff
@@ -0,0 +1,72 @@
+import { ErrorExceptionType } from "../constants"
+import { isUUID } from "../utils"
+
+export type AssignmentRequestDto = {
```

Comment:
I think this DTO is too generic, I don't like this "one dto for all" approach. It would be more expressive to create separate ones for each operation.

Reply:
This one. I completely agree. I spent about 1 hour thinking should I do "one to rule them all" or create separate ones for specific operations. I'm just wondering how do you know if an operation needs a separate DTO. For example, operation A and operation B are identical in validation and returning the result but they are inside different services. Would you still create a separate DTO for each one or one to cover them both? I kinda struggle with this concept...

Reply:
Good question. I think I'd go for creating separate DTOs for both in order to express that they serve different purposes even though they look and act the same.


Ex.:

I think you could try creating an abstraction that encapsulates the common things of A and B (in this case everything) but still express that they are different by using different names.
This can help you to reduce code duplication for the time being. Once the 2 things start to diverge, or you discover that this is just [incidental duplication](https://anthonysciamanna.com/2018/07/28/the-dry-principle-and-incidental-duplication.html) you could get rid of the abstraction and actually create 2 different types/implementations.
This way your code already expresses that A and B serve different purposes, and you can save some refactoring time once they start to diverge. 

```ts
type BaseAssignmentDTO = {...};

export type CreateAssignmentDTO = BaseAssignmentDTO; // used by Assignments
export type CreateStudentAssingmentDTO = BaseAssignmentDTO // used by StudentAssignments
```

Reply:
Good question. I think I'd go for creating separate DTOs for both in order to express that they serve different purposes even though they look and act the same.


Ex.:

I think you could try creating an abstraction that encapsulates the common things of A and B (in this case everything) but still express that they are different by using different names.
This can help you to reduce code duplication for the time being. Once the 2 things start to diverge, or you discover that this is just [incidental duplication](https://anthonysciamanna.com/2018/07/28/the-dry-principle-and-incidental-duplication.html) you could get rid of the abstraction and actually create 2 different types/implementations.
This way your code already expresses that A and B serve different purposes, and you can save some refactoring time once they start to diverge. 

```ts
type BaseAssignmentDTO = {...};

export type CreateAssignmentDTO = BaseAssignmentDTO; // used by Assignments
export type CreateStudentAssingmentDTO = BaseAssignmentDTO // used by StudentAssignments
```